muspinsim/input/larkeval.py

```diff
@@ -192,13 +194,15 @@ def evaluate(self, **variables):
         vset = set(variables.keys())
         if len(self.variables - vset) > 0:
             raise LarkExpressionError(
-                "Some necessary variables have not been "
-                "defined when evaluating LarkExpression"
+                "Some necessary variable(s) {0} have not been "
+                "defined when evaluating LarkExpression".format(self.variables - vset)
             )
         elif len(vset - self._all_variables) > 0:
             raise LarkExpressionError(
-                "Some invalid variables have been "
-                "defined when evaluating LarkExpression"
+                "Some invalid variable(s) {0} have been "
+                "defined when evaluating LarkExpression".format(
+                    vset - self._all_variables
+                )
```

Comment:
I've been trying to figure out how to trigger these errors but it's difficult to do(!). Do you know what sets them off?

I had success if I included both `fitting_variables` and `experiment` keywords, but in that case this error message wasn't very helpful as the blocks were both fine in and of themselves, it was the inter-incompatibility that broke things.

Reply:
could you provide the input file you used?

Reply:
Did you get anywhere with understanding how this error is triggered?

Reply:
So this error cannot b triggered by a user if they just run muspinsim via cmd:
`muspinsim input.in`

or they do something like

```
p = MuSpinInput(open('fitting.in'))
p2 = FittingRunner(p)
run = p2.run()
p2.write_report()
```

The error can only be triggered during `p2.run()` - when we run the fitting algorithm. Here, we iteratively call ExperimentRunner with calculated values for each fitting variable. 

fitting,py (104-105)
```
vardict = dict(zip(self._xnames, self._x))
self._runner = ExperimentRunner(self._input, variables=vardict)
```

If for any reason the dictionary of variables calculated `vardict` does not contain an expected value or contains a value which shouldn't exist then these errors are raised

So really, the only way this can be triggered is if the user wants to define their own fitting algorithm instead of calling `p2.run()` or something like that

Reply:
For me, the error is triggered if I use the input file above for muspinsim - `muspinsim.input.larkeval.LarkExpressionError: Some invalid variable(s) {'g'} have been defined when evaluating LarkExpression`.

I'm trying to figure out if specifying an `experiment` breaks the fitting somehow, or if certain parameters can't be used together for physical reasons. I don't understand why it's triggering this specific error.

Reply:
I found the source of the bug - misunderstanding on my part

`experiment` keyword breaks because the code uses a recursive call to parse the keywords generated.
https://github.com/muon-spectroscopy-computational-project/muspinsim/blob/main/muspinsim/input/input.py#L116-L117 
```
mock_i = MuSpinInput(StringIO(_exp_defaults[exptype[0]]))
self._keywords.update(mock_i._keywords)
```
when it does so, it fails to pick up fitting variables and causes an error.

Reply:
I fixed this error - but now we get another
```
IndexError: invalid index to scalar variable.
```
I think this is due to improper inputs rather than a bug in the code

Reply:
> I think this is due to improper inputs rather than a bug in the code

Seems like it - it works if `experiment` is `zero_field` but not `alc`. Thanks for the fix, I'm happy with this now.